### PR TITLE
GraphQL association field security fixes + @ApiProperty security support

### PR DESCRIPTION
**Pull request by DavidBennettUK**
| Q             | A
| ------------- | ---
| Branch?       | 2.6
| Bug fix?      | yes
| New feature?  | no
| Deprecations? |no
| Tickets       | https://github.com/api-platform/core/issues/4169
| License       | MIT
| Doc PR        | N/A

The security attribute was added to `ApiProperty` here: https://github.com/api-platform/core/pull/3503 which lets you add admin-only properties etc. (https://api-platform.com/docs/core/security/)

However, the GraphQL `ItemNormalizer` wasn't updated to inject the new `ResourceAccessCheckerInterface` dependency into the parent, and would fail silently. This PR adds this missing injection.

This should probably be a separate PR as it's a different issue that's been uncovered now this is fixed, but if you add security to a non-null GraphQL property then you will get a `Cannot return null for non-nullable field` error if trying to request a field you don't have access to. I'm not sure if it would be best if an access denied error was given instead, update the way the schema is built to detect that the field is nullable if security is present, or leave it to the user to make the field nullable.

### CODE REVIEW

**Comment by alanpoulain** [comment]
Hello, thanks for the PR.

You can find a related discussion here: https://github.com/api-platform/core/pull/3227#issuecomment-556013267
And an issue for adding a `visibility` attribute here (complementary with the `security` one): https://github.com/api-platform/core/issues/3279

**Comment by alanpoulain** [comment]
I think we cannot merge this PR without solving the underlying issue.

**Comment by DavidBennettUK** [comment]
Thanks, I'd missed that discussion. I agree that the field should be marked as nullable if the security attribute is present. I've updated the PR with a suggestion for how this can be done, which seems to work well in my project - working how I'd expect. I'm not super familiar with the internals of this project, so this may not be the best way. I think the visibility attribute might be a separate thing, as I'd still want the property in the schema/docs for use with generating TypeScript types.

**Comment by alanpoulain** [comment]
Yes the `visibility` attribute should be another PR indeed.
WDYT of throwing an exception if the field is not nullable? I think it will be less "magic", don't you think?

**Comment by DavidBennettUK** [comment]
I agree, it's definitely a bit magic. It does make sense though, otherwise an invalid schema is generated if user forgets to make the field nullable. If there's a nice way to set a field's graph type as nullable I'd be for throwing an exception instead. However, unless I've missed something, I haven't seen a way to make a field's graph type nullable, without changing the Doctrine `@ORM\Column` to nullable (wouldn't want to change the database schema, as we shouldn't need to) - with the exception of writing a `TypeConverter`, which is a bit cumbersome. Do you know of a way?

Separately, I've encountered an issue trying to apply security to a Doctrine collection field. If the authorization fails, the collection field isn't present in the data returned by the graph [ItemNormalizer](https://github.com/api-platform/core/blob/2.6/src/GraphQl/Serializer/ItemNormalizer.php#L84) - which is correct. However, the collection field is still in the response due to the [CollectionResolver](https://github.com/api-platform/core/blob/2.6/src/GraphQl/Resolver/Factory/CollectionResolverFactory.php#L90) being run, because it was set up by in [FieldsBuilder](https://github.com/api-platform/core/blob/2.6/src/GraphQl/Type/FieldsBuilder.php#L314) beforehand. This is getting a bit much for my knowledge of the project, so I'm not sure how to fix this. Any pointers?

I should probably open a separate issue for this, but I've noticed that the security stage in CollectionResolver uses the `item_query` security of the child type in the collection, with a `Paginator` as the object. I'm not sure if this is the correct behaviour, as `item_query` security would expect an instance of the entity, not a paginator. More [here](https://github.com/api-platform/core/issues/3535#issuecomment-723212145).

For the time being, is there another way to secure graph fields? I'm aware that it's possible to add dynamic serializer groups, but the issue there is that the dynamic fields aren't added to the generated documentation/schema - which means that using the schema for TypeScript type generation etc. won't work.

**Comment by alanpoulain** [comment]
> Do you know of a way?

You're right, the Doctrine property type extractor has the highest priority. We should definitely solve this issue. I'm not sure how though so maybe the right way is like you did in this PR.

> Any pointers?

This one is tricky indeed. I think we can check if the field is not present in the `$source` array. If it's the case, it means that a security has been applied and we can return directly. WDYT?

> I should probably open a separate issue for this, but I've noticed that the security stage in CollectionResolver uses the `item_query` security of the child type in the collection, with a `Paginator` as the object. I'm not sure if this is the correct behaviour, as `item_query` security would expect an instance of the entity, not a paginator. More [here](https://github.com/api-platform/core/issues/3535#issuecomment-723212145).

Seems like a bug alright. Maybe we should always use `collection_query`. I need to think about it.

> For the time being, is there another way to secure graph fields?

Not for the moment :slightly_smiling_face:

**Comment by DavidBennettUK** [comment]
Sorry for the late reply, I haven't had a chance to revisit this yet. I've opened up an issue (https://github.com/api-platform/core/issues/4169) regarding the security issues to see if this is worth pursuing.

**Comment by DavidBennettUK** [comment]
@alanpoulain I've implemented fixes for the issues raised above (more detail in https://github.com/api-platform/core/issues/4169) + finished implementing `@ApiProperty` security using your suggestion 👍 It all seems to be working well now - security is running how I expect on relations, and fields (including relations) can now be secured. I've updated the PHPUnit tests and added a set of Behat tests to cover the usage scenarios.

Think this is ready to go if you think it looks good? Not sure if this counts as a breaking change or not - would you want me to target another branch?

File: src/GraphQl/Type/FieldsBuilder.php
Comment (alanpoulain) [inline src/GraphQl/Type/FieldsBuilder.php:501]: I would make it first since it's the most understandable and the less costly.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php:62]: I think a comment would be helpful to understand what is going on here. Like:
If authorization has failed for a relation field, the field is not present in the source: null can be returned directly.

File: features/graphql/authorization.feature
Comment (alanpoulain) [inline features/graphql/authorization.feature:113]:
```suggestion
    And I send the following GraphQL request:
```

File: features/graphql/authorization.feature
Comment (alanpoulain) [inline features/graphql/authorization.feature:130]:
```suggestion
    And the JSON node "data.securedDummy.relatedDummies" should have 1 element
```

File: features/graphql/authorization.feature
Comment (alanpoulain) [inline features/graphql/authorization.feature:134]:
```suggestion
    And I send the following GraphQL request:
```

File: features/graphql/authorization.feature
Comment (alanpoulain) [inline features/graphql/authorization.feature:153]:
```suggestion
    And I send the following GraphQL request:
```

File: features/graphql/authorization.feature
Comment (alanpoulain) [inline features/graphql/authorization.feature:174]:
```suggestion
    And I send the following GraphQL request:
```

File: tests/Fixtures/TestBundle/Document/SecuredDummy.php
Comment (alanpoulain) [inline tests/Fixtures/TestBundle/Document/SecuredDummy.php:94]:
```suggestion
     * @var Collection<RelatedDummy>
```

File: tests/Fixtures/TestBundle/Entity/SecuredDummy.php
Comment (alanpoulain) [inline tests/Fixtures/TestBundle/Entity/SecuredDummy.php:95]:
```suggestion
     * @var Collection<RelatedDummy>
```

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php:91]: I think it's dangerous to do this. If the developer has added security in the related resource, they think it will be always applied in all contexts, don't you think?
Only applying the security at the top level means that if a user has access to a resource, it can access to secured resources by traversing the relations (if the relations have not been secured with your change of course).
I think the better approach is to change the operation name to be always `collection_query`.

File: tests/GraphQl/Resolver/Factory/CollectionResolverFactoryTest.php
Comment (alanpoulain) [inline tests/GraphQl/Resolver/Factory/CollectionResolverFactoryTest.php:113]: We should add a test when the source is not null but the field name is not present as well.

**Review by alanpoulain — CHANGES_REQUESTED** [review]
Since there are a lot changes, I think it would be safer to target `main` indeed.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php
Comment (DavidBennettUK) [reply src/GraphQl/Resolver/Factory/CollectionResolverFactory.php:92]: I do see your point, and would certainly want to update the documentation to make it clearer if we do go this route.
That said, I see the `ApiResource` GraphQL `item_query`, `collection_query`, etc. as definining the actual GraphQL entrypoints i.e. the top-level queries.
Association collections are different collections to the main resource collection - they're filtered. The `collection_query` in `ApiResource` exposes the entire collection. I don't think it makes sense to use the same security on associations, as the requirements may be different. In my example (in https://github.com/api-platform/core/issues/4169) you would want admins to see the whole collection of documents, and not let normal users see them all, instead letting users only see their own documents via the `user.documents` association. Using the `collection_query` security on the association would mean only admins could ever see it.
You may also not want to have a `collection_query` exposing the whole collection as an endpoint, but still want to have associations. Relying on there being a `collection_query` present wouldn't work in this case.
Unless I've found a bug, in 2.6.3 at least, this PR makes collection security work the same way as single associations already do. `item_query` security from the association resource is not run on associations - meaning they are already exposed with no way to protect them. This is covered by the following two tests I've added:
- https://github.com/api-platform/core/pull/4143/files#diff-14861a046eb66cb65d4007864a6fbe8d08b64a6cd05a61f21d5cab03b7eb7525R193 - shows that it's not possible for a user to access an `item_query` secured resource directly
- https://github.com/api-platform/core/pull/4143/files#diff-14861a046eb66cb65d4007864a6fbe8d08b64a6cd05a61f21d5cab03b7eb7525R254 - shows that it's possible to access the `item_query` secured resource via a association. The `item_query` security is never run.

I hope that all makes sense!

**Comment by alanpoulain** [comment]
I have changed the targeted branch to `main`, could you rebase your `2.6` branch on `main` (for future PR you should create a new branch)?
I have no right to push on your branch, you should allow maintainers to push on it, it makes things easier for us:
image

**Comment by DavidBennettUK** [comment]
@alanpoulain thanks, I've rebased onto `main`. It looks like because I created the fork in an organisation, it's not possible to allow edits to maintainers except inviting them individually (https://github.com/isaacs/github/issues/1681) - I'll remember for next time. I've also added some entries to the changelog to cover the changes in this PR.

File: src/Serializer/AbstractItemNormalizer.php
Comment (alanpoulain) [inline src/Serializer/AbstractItemNormalizer.php:255]: Is this change necessary? Could you denormalize and update / create a secured field without the right to do so before?

File: src/Serializer/AbstractItemNormalizer.php
Comment (alanpoulain) [inline src/Serializer/AbstractItemNormalizer.php:387]: You cannot change the typehint, it's a BC break.

File: src/Serializer/AbstractItemNormalizer.php
Comment (alanpoulain) [inline src/Serializer/AbstractItemNormalizer.php:399]: It seems it can also be a `class-string`.

File: src/Serializer/AbstractItemNormalizer.php
Comment (alanpoulain) [inline src/Serializer/AbstractItemNormalizer.php:393]: Isn't it dangerous to do so? We could add a security without an object, isn't it?

**Comment by alanpoulain** [comment]
For the change "Security: Fix `ApiProperty` security being passed a null object for updates.", please do it in another PR, it will be easier to discuss about it and I think it's not really related to this PR.

**Comment by DavidBennettUK** [comment]
@alanpoulain that's fair, I did consider splitting it up, as it fixes an issue not just for GraphQL but all. I've removed it from this PR and will open another.

**Review by alanpoulain — APPROVED** [review]

**Comment by alanpoulain** [comment]
This is some great work! Thank you very much @DavidBennettUK.
Can I ask you to write some documentation about it? I think it would be very helpful.

**Comment by DavidBennettUK** [comment]
Thanks @alanpoulain - happy to be able to contribute to the project! I appreciate your thorough feedback and help. I'll get some documentation written up and put in a PR in the next few days - I spent a bit longer working on the other PR (https://github.com/api-platform/core/pull/4184) than I expected, but that's ready to be looked at now at least.

**Comment by DavidBennettUK** [comment]
@alanpoulain Got the documention PR ready 👍  https://github.com/api-platform/docs/pull/1338

**Comment by webda2l** [comment]
Thanks for this PR and #4184!

@alanpoulain It works great on `dev-main`, but any reason that it's to not already part of a 2.6 realease? If yes, any ETA to have this code on a stable (2.7 or 3.0)? Thank you!

**Comment by alanpoulain** [comment]
It will be available for `2.7` but we don't have any ETA.

**Comment by carlobeltrame** [comment]
Since this PR is relevant to security and there is no definite release date for 2.7, should we backport this to 2.6.x?

**Pull request merged**
